docs/viplab3.0/versioning.md

```diff
@@ -0,0 +1,62 @@
+# JSON Object Versioning Rules
+
+* Status: proposed
+* Deciders: Leon Kiefer, Per Pascal Seeland
+* Date: 2020-05-14
+
+Technical Story: https://github.com/VirtualProgrammingLab/documentation/issues/1
+
+## Context and Problem Statement
+
+The ViPLab Projects defines a set of JSON-objects, which are used in the different components of the project. Since ViPLab itself does not 
+store any data, these objects have to be stored by third party systems. These will be, depending on the context they are used in either LMS (like ILIAS)
+or data repositories (like dataverse). There needs to be a way to identify the version of the different objects stored by the third party system to 
+allow e.g. upgrading/migrating the different components. 
+
+## Decision Drivers
+
+* Provide a way to identify the version of the objects and also provide a path for e.g. upgrading the API endpoints used by those systems.
+
+## Considered Options
+
+* Version all endpoints, but not the objects
+* Version all objects, but not the endpoints
+* Version parts of the objects and all of the endpoints.
+* Version all objects and all of the endpoints.
+
+## Decision Outcome
+
+It has been decided to version all long living objects, and all endpoints. The reason for requiring all objects to have version tags is for consistency
+reasons. Endpoints will only accept objects versions of one major version. Breaking changes will require to define a new version. Incremental 
+updates, which only extend the object are allowed. All systems are expected to handle these incremental updates gracefully. This means that a
+version 3.1.x message returned by the api must be accept by a 3.0.x client. The additional information contained by the api may be discarded. The same
+rules apply when a 3.1.x client talks to a 3.0.x api. The api will process the message according to the 3.0.x rules and extra information might be discarded.
+
+The client has to determine the correct endpoint versions to use. When the client will use a ComputationTemplate with version 3.x.x the corresponding
+websocket api will be ws://example.com/v3/compuation. All other messages used in any further communication between the websocket-api and the client
+are also expected to be version 3.x.x. There is no possiblity to mix different versions. If a client wants to use a v4 api, he has to reconnect to the correct
+endpoint at ws://example.com/v4/compuation
+
+The objects have been devided into two groups:
+- Long living object: These objects are expected to be stored by clients. These objects are ComputationTemplate, the ComputationTask and the ComputationResult
+- Action Wrappers: Computation and Result
```

Comment:
[Create Computation](https://github.com/VirtualProgrammingLab/viplab-websocket-api/blob/master/schema/create-computation-schema.json) and [Computation](https://github.com/VirtualProgrammingLab/viplab-websocket-api/blob/master/schema/computation-schema.json) are short lived interaction objects/wrappers. They only exist for the purpose of communicating information in the websocket connection and as mentioned are versionen by the API endpoint.

But Result as it is currently discussed in VirtualProgrammingLab/viplab-websocket-api#14 is not only an object which is created for the websocket API. It is created by the backends and send via the websocket API to the clients, so it must have a version and can not be versioned with the endpoint.

Reply:
For Result, I think we have a naming issue. The concept described in  VirtualProgrammingLab/viplab-websocket-api#14 does belong to the ComputationResult. And those would, as being long living objects, have an version.
The Result I mentioned in the Action Wrappers is actually the message type that gets send to the client from the api.
Maybe, we need to define three types of objects. Long living (they get versioned), message types(unversioned, like Create Computation and Result) and Wrappers (unversioned, Computation). The unversioned types are only specified at an endpoint level, and require the endpoint version.

Reply:
Ok, yes